### PR TITLE
[WIP] On-the-fly transformation

### PR DESCRIPTION
**Pull request by joegomes**
This PR adds:
- An implementation of `TensorflowMultiTaskFitTransformRegressor`.  Some featurizers (such as `CoulombMatrix`) produce feature vectors that are not invariant to atom index permutation.  Models that use these features are improved when the atom indices for a given molecule are randomized at fit/predict time.  This model allows for arbitrary `fit_transformers` to be applied to the feature vector during training and testing.  Currently only supports `CoulombMatrix` featurizer but can be generalized to other featurizers in the future.  Fit transformation in `predict()` is WIP.
- An implementation of `CoulombRandomizationFitTransformer` and `NormalizationFitTransformer`.  These are the first two examples of fit transformers that can be applied with the `TensorflowMultiTaskFitTransformerRegressor`.  `BinarizationFitTransformer` is WIP.
- An update to the GDB7 MT-NN benchmark.  So far, train/test scores have improved from 86.1/92.2 to 56.5/69.2 kcal/mol MAE.  These numbers should improve once `predict()` fit transformation and `BinarizationFitTransformer` are improved.  Implementing a learning rate decay schedule and training for 24 hours should get us close to the QMachine benchmark.

### CODE REVIEW

**Comment by rbharath** [comment]
@joegomes Very cool! Would you mind updating README.md with the new numbers? (It's kind of nice to have a public "scoreboard" to track where we're at).

Let me know once this PR is ready to merge.

**Comment by joegomes** [comment]
@rbharath OK, I've added the new numbers to the README.md.  Index: 18.3/172.0 kcal/mol MAE; random: 44.3/59.1 kcal/mol MAE.  I'm going to finish the learning rate schedule implementation later this evening, and assuming tests pass, this should be ready to merge in then.

**Comment by joegomes** [comment]
@miaecle Do you think it's possible to implement the 5-fold cross-validation used in the reference paper into the gdb7_tf_model.py after this is merged?  We can switch to user-specified features and use their dataset file if that makes it easier.  I think that is all that is missing from reproducing the results.

**Comment by miaecle** [comment]
@joegomes Thanks a lot for getting these modifications in! For reproducing their results, I have actually imported an `IndiceSplitter` into `gdb7_datasets.py,` so it should be very straightforward to use their splitting indices(in the file `gdb7_splits.csv`) to train and test by setting `split="indice"` in `load_gdb7` function. A little modification is needed to implement cross validation(running 5 times with different indices), but will be very easy. Also, if we are going to completely reproduce the results, maybe directly using their features(X) will be of some help. Guess we can do both?

**Comment by joegomes** [comment]
@rbharath @miaecle I've implemented the cross-validation scheme from the paper into the gdb7 benchmark, as well as adding in the learning rate schedule.  Here's some numbers I get training for 750 epoch (30 min):

computed_metrics: [4.583430741213796]
computed_metrics: [0.99911767425370157]
Train scores [kcal/mol]
{'pearson_r2_score': 0.99911767425370157, 'mean_absolute_error': 4.583430741213796}
computed_metrics: [5.5591552199112098]
computed_metrics: [0.99830352306686321]
Test scores [kcal/mol]
{'pearson_r2_score': 0.99830352306686321, 'mean_absolute_error': 5.5591552199112098}

I've also updated the README.md with the results of a 50 epoch optimization.  I think these split are technically user-defined, so I added a new row.  I haven't re-run the index and random splits yet with the new `TensorflowCoulombMatrixRegressor` model nor have I fully-integrated this into the test suite.  I think the biggest gain in performance came from using their splits, we should read more about their stratified split technique as it might be useful elsewhere.  This is ready to merge when you get a chance to go over it!

**Comment by rbharath** [comment]
@joegomes Thanks for getting this implemented! It's very interesting that the biggest boost comes from using their splits. Suggests that overfit is a serious problem with the numbers reported in papers... The random splits are much worse (indicating that real world uses of these predictors will likely have terrible performance). Something worth discussing more offline.

Before we merge in, would you mind adding unit tests for `CoulombFitTransformer`, `CoulombRandomTransformer` and the other new classes? This code is complicated enough that I want to make sure it's tied into the test suite (so refactoring won't break it).

**Comment by joegomes** [comment]
@rbharath Sounds good, I will add some tests for the new classes this PR adds and let you know when I'm ready.  I will probably get around to it in the next day or so.

**Comment by rbharath** [comment]
@joegomes Since we have a large backlog of open PRs, I'm inclined just to go ahead and merge this as-is (with understanding that tests will be added later). Would that cause any issues?

**Comment by joegomes** [comment]
@rbharath I will have some time tonight/tomorrow to write up the unit tests, documentation, and example code for the new classes.  Let's hold off on merging the PR until the end of the weekend.  I'll be doing my best to maintain best practices moving forward =)

**Comment by joegomes** [comment]
@rbharath Closing and re-opening a clean PR.
